OpenRA.Test/OpenRA.Game/ShapeTest.cs

```diff
@@ -0,0 +1,96 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.IO;
+using NUnit.Framework;
+using OpenRA.Mods.Common.HitShapes;
+
+namespace OpenRA.Test
+{
+	[TestFixture]
+	public class ShapeTest
+	{
+		IHitShape shape;
+
+		[TestCase(TestName = "CircleShape reports accurate distance")]
+		public void Circle()
+		{
+			shape = new CircleShape(new WDist(1234));
+			shape.Initialize();
+
+			Assert.That(shape.DistanceFromEdge(new WVec(100, 100, 0)).Length,
+				Is.EqualTo(0));
+
+			Assert.That(shape.DistanceFromEdge(new WVec(1000, 0, 0)).Length,
+				Is.EqualTo(0));
+
+			Assert.That(shape.DistanceFromEdge(new WVec(2000, 2000, 0)).Length,
+				Is.InRange(1593, 1595));
```

Comment:
It is deterministic with well defined rules: you will always get the same result.